lib/rmg/TreasurePlacer.cpp

```diff
@@ -0,0 +1,795 @@
+			
+			oi.setTemplate(Obj::SEER_HUT, randomAppearance, zone.getTerrainType());
+			oi.value = generator.getConfig().questValues[i];
+			oi.probability = 10;
+			
+			oi.generateObject = [i, randomAppearance, this, generateArtInfo]() -> CGObjectInstance *
+			{
+				auto factory = VLC->objtypeh->getHandlerFor(Obj::SEER_HUT, randomAppearance);
+				auto obj = (CGSeerHut *) factory->create(ObjectTemplate());
+				
+				obj->rewardType = CGSeerHut::EXPERIENCE;
+				obj->rID = 0; //unitialized?
+				obj->rVal = generator.getConfig().questRewardValues[i];
+				
+				obj->quest->missionType = CQuest::MISSION_ART;
+				ArtifactID artid = *RandomGeneratorUtil::nextItem(generator.getQuestArtsRemaning(), generator.rand);
+				obj->quest->m5arts.push_back(artid);
+				obj->quest->lastDay = -1;
+				obj->quest->isCustomFirst = obj->quest->isCustomNext = obj->quest->isCustomComplete = false;
+				
+				generator.banQuestArt(artid);
+				
+				this->questArtZone->getModificator<TreasurePlacer>()->possibleObjects.push_back(generateArtInfo(artid));
+				
+				return obj;
+			};
+			
+			possibleObjects.push_back(oi);
+			
+			oi.generateObject = [i, randomAppearance, this, generateArtInfo]() -> CGObjectInstance *
+			{
+				auto factory = VLC->objtypeh->getHandlerFor(Obj::SEER_HUT, randomAppearance);
+				auto obj = (CGSeerHut *) factory->create(ObjectTemplate());
+				obj->rewardType = CGSeerHut::RESOURCES;
+				obj->rID = Res::GOLD;
+				obj->rVal = generator.getConfig().questRewardValues[i];
+				
+				obj->quest->missionType = CQuest::MISSION_ART;
+				ArtifactID artid = *RandomGeneratorUtil::nextItem(generator.getQuestArtsRemaning(), generator.rand);
+				obj->quest->m5arts.push_back(artid);
+				obj->quest->lastDay = -1;
+				obj->quest->isCustomFirst = obj->quest->isCustomNext = obj->quest->isCustomComplete = false;
+				
+				generator.banQuestArt(artid);
+				
+				this->questArtZone->getModificator<TreasurePlacer>()->possibleObjects.push_back(generateArtInfo(artid));
+				
+				return obj;
+			};
+			
+			possibleObjects.push_back(oi);
+		}
+	}
+}
+
+bool TreasurePlacer::isGuardNeededForTreasure(int value)
+{
+	return zone.getType() != ETemplateZoneType::WATER && value > minGuardedValue;
+}
+
+std::vector<ObjectInfo*> TreasurePlacer::prepareTreasurePile(const CTreasureInfo& treasureInfo)
+{
+	std::vector<ObjectInfo*> objectInfos;
+	int maxValue = treasureInfo.max;
+	int minValue = treasureInfo.min;
+	
+	const ui32 desiredValue = generator.rand.nextInt(minValue, maxValue);
+	
+	int currentValue = 0;
+	bool hasLargeObject = false;
+	while(currentValue <= (int)desiredValue - 100) //no objects with value below 100 are available
+	{
+		auto oi = getRandomObject(desiredValue, currentValue, maxValue, !hasLargeObject);
+		if(!oi) //fail
+			break;
+		
+		if(oi->templ.isVisitableFromTop())
+		{
+			objectInfos.push_back(oi);
+		}
+		else
+		{
+			if(hasLargeObject)
+			{
+				//only one large object per treasure pile
+				continue;
```

Comment:
Can we get into endless loop here?

Reply:
Good point. If all objects are large we may get into infinite cycle. Will thing how to fix

Reply:
getRandomObject does same check, probably I can get rid of this condition